Modify storage of Layers view queries and preferences

aldebaran30701 opened the pull request:
<!--

### Requirements

* Filling out the template is required. Any pull request that does not include
enough information to be reviewed in a timely manner may be closed at the
maintainers' discretion.
* Follow the check list items defined by https://github.com/constellation-app/constellation/blob/master/CONTRIBUTING.md#pull-requests
* All new code requires unit tests to ensure they work as expected and will
continue to work as new code is added in the future (regression testing).
* Have you read Constellation's Code of Conduct? By filing an issue, you are
expected to comply with it, including treating everyone with respect:
https://github.com/constellation-app/constellation/blob/master/CODE_OF_CONDUCT.md

-->

### Description of the Change
This change fixes #589 and implements a future on saving of queries to ensure only one selection is needed to have a layer become activated.

LayerQueries list which used to be stored inside of StoreGraph was causing issues as the DualGraph managed two instances of StoreGraphs which were out of sync - causing odd behaviour like changing of attribute values. 
The fix stores the list of queries inside a GraphAttribute Object. The same migration process was done to LayerPreferences to fix an issue which arose from the above implementation. 
<!--

We must be able to understand the design of your change from this description.
If we can't get a good idea of what the code will be doing from the description
here, the pull request may be closed at the maintainers' discretion. Keep in
mind that the maintainer reviewing this PR may not be familiar with or have
worked with the code here recently, so please walk us through the concepts.

-->

### Alternate Designs
Trigger a commit whenever one StoreGraph layerQueries list was updated, to force the other StoreGraph within the DualGraph to sync with the updated one. I did not get far into this idea before running into problems.
<!--

Explain what other alternates were considered and why the proposed version was
selected.

-->

### Why Should This Be In Core?
Bug Fix - enhances the usability of layers view
<!--

Explain why this functionality should be in Constellation Core as opposed to a
different module suite. Note that this question is more applicable when adding
new functionality. If this change is a minor update to an exising file then it
is understood that this change has to be to this module suite and a response
and therefore a response to this question is not required.

-->

### Benefits
layers now work on first click, and react normally because the two graphs are in sync
<!-- What benefits will be realized by the code change? -->

### Possible Drawbacks
None realised
<!-- What are the possible side-effects or negative impacts of the code change? -->

### Verification Process
Testing Queries
Add color and other attribute queries to each layer, verify correct results
2.	selecting a section of the graph, right click context menu -> add to a vacant layer
a.	Ensure the selection is correct with what is added to the layer
3.	Select more elements and test remove functionality
a.	Ensure the elements are removed from that layer
4.	Test layers with short or long names for visuals and any errors.
Testing Draw mode
1.	Create a new graph
a.	Add a few elements and transactions on
i.	Should appear visible
b.	Open layers view
i.	Select layer 2 only
ii.	Add elements (Vertices and Transactions)
iii.	Elements should only appear on layer 2
<!--

What process did you follow to verify that your change has the desired effects?

- How did you verify that all new functionality works as expected?
- How did you verify that all changed functionality works as expected?
- How did you verify that the change has not introduced any regressions?

Describe the actions you performed (e.g., buttons you clicked, text you typed,
commands you ran, etc.), and describe the results you observed.

-->

### Applicable Issues
#589 
<!-- Link any applicable issues here -->

arcturus2 commented:
@aldebaran30701 so Travis is failing. Can you please check that you have all the latest changes from master?

aldebaran30701 commented:
> @aldebaran30701 so Travis is failing. Can you please check that you have all the latest changes from master?

I did update from master just before submitting the PR, I've just pulled and merged everything currently on master - hopefully this fixes it. I'm just triple checking and it builds locally for myself.

antares1470 commented on CoreGraphFramework/src/au/gov/asd/tac/constellation/graph/StoreGraph.java:
```diff
@@ -1969,18 +1953,24 @@ public void clearValue(final int attribute, final int id) {
     private void recalculateLayerVisibilities() {
-        for (int i = 0; i < layerQueries.size(); i++) {
+        List<String> queries = getQueries();
+        List<Byte> preferences = getLayerPreferences();
```
above 2 could be made final

antares1470 commented on CoreGraphFramework/src/au/gov/asd/tac/constellation/graph/StoreGraph.java:
```diff
@@ -1948,6 +1908,30 @@ public void clearValue(final int attribute, final int id) {
+        if (queriesId == Graph.NOT_FOUND) {
+            return Collections.emptyList();
+        }
+        Object queriesObject = getObjectValue(queriesId, 0);
```
final

antares1470 commented on CoreGraphFramework/src/au/gov/asd/tac/constellation/graph/StoreGraph.java:
```diff
@@ -1948,6 +1908,30 @@ public void clearValue(final int attribute, final int id) {
+        if (preferencesId == Graph.NOT_FOUND) {
+            return Collections.emptyList();
+        }
+        Object preferencesObject = getObjectValue(preferencesId, 0);
```
final

antares1470 commented on CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/LayersViewController.java:
```diff
@@ -67,36 +70,27 @@ public void edit(final GraphWriteMethods graph, final PluginInteraction interact
-        PluginExecution.withPlugin(new UpdateGraphBitmaskPlugin(newBitmask))
+
+        Future<?> layerQueryFuture = PluginExecution.withPlugin(new UpdateGraphQueriesPlugin(layerQueries))
                 .executeLater(GraphManager.getDefault().getActiveGraph());
```
Is the active graph required here? Just wondering about executing this all standalone (e.g. in a unit test)

aldebaran30701 replied:
I think I get where you are coming from - in the case where there is no active graph, or the graph is closed? 
I'm not too aware of other ways - I've seen this way as a standard throughout the project - I could be using it in the wrong context though. @arcturus2 Would a null check here or within the plugin to be executed be necessary?

arcturus2 replied:
> I've seen this way as a standard throughout the project - I could be using it in the wrong context though.

I agree @aldebaran30701 it used throughout the code base and context does matter. In general a view should be fine to make the `getActiveGraph()` call though checking for null here would be required.

I've opted to not do a null check in `executeLater` deliberately because situations like this can go unnoticed.

aldebaran30701 replied:
In an attempt to not let situations like this go unnoticed should we add a warning log or something if the current active graph is null?
If the current active graph is null, nothing further should be done correct?

arcturus2 replied:
Yeah @aldebaran30701 I think logging it is a great idea.

So on the topic of things silently passing/failing I commented https://github.com/constellation-app/constellation/pull/602#discussion_r447458005 that it turns out the plugin is not run in the unit test; a good example of how things get missed. Doing the null check before running `executeLater` to my mind makes it more obvious when the plugin runs.

arcturus2 replied:
>  If the current active graph is null, nothing further should be done correct?

Correct. In `executeNow()` it looks like there is a check for `null` but not in `executeLater`. One silently fails and the other throws an exception!

Perhaps a `null` check and logging it would be fine. This would be a change in the `DefaultPluginEnvironment` etc and a separate PR I guess. Want me to do it or are you ok to @aldebaran30701 ?

aldebaran30701 replied:
Just to reiterate - This code block does not change, but a separate Pr which has a null check and error logging within `DefaultPluginEnvironment ` is required correct?

arcturus2 replied:
Yeah @aldebaran30701 and this is largely to keep your PR focused on the layers code and also because the plugin framework not dealing with nulls (on purpose) is a different problem.

antares1470 commented on CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/LayersViewController.java:
```diff
@@ -67,36 +70,27 @@ public void edit(final GraphWriteMethods graph, final PluginInteraction interact
-        }
-        PluginExecution.withPlugin(new UpdateGraphQueriesPlugin(layerQueries))
+        PluginExecution.withPlugin(new UpdateGraphBitmaskPlugin(newBitmask)).waitingFor(layerQueryFuture)
                 .executeLater(GraphManager.getDefault().getActiveGraph());
```
same here

arcturus2 commented on CoreLayersView/test/unit/src/au/gov/asd/tac/constellation/views/layers/utilities/LayersDualGraphSyncNGTest.java:
```diff
@@ -0,0 +1,318 @@
+                final int maxTransactionsId = VisualConcept.GraphAttribute.MAX_TRANSACTIONS.ensure(graph);
+                graph.setIntValue(maxTransactionsId, 0, 9);
+            }
+        }).executeNow(GraphManager.getDefault().getActiveGraph());
```
@aldebaran30701 this is similar to the problem I am presently looking at. In a unit test `getActiveGraph()` will be `null` so you will find that this plugin isn't executed. In the `edit()` if you print something out I think you will find that it isn't appearing.

The fix would be to pass in `graph`.

arcturus2 replied:
@aldebaran30701 I updated it to use `graph` and it passed - so executing `graph.setIntValue(maxTransactionsId, 0, 9);` had no net effect. Do you think a revisit of the unit test or layers code is required?

aldebaran30701 replied:
That would explain why the graph out of sync issue was not reproducible through the unit tests. I assumed it was a product of running it on a single thread.

aldebaran30701 replied:
So I have also had a play around with this, and it seems that because `getActiveGraph()` returns null, and graph is the same instance of `StoreGraph`, I cannot get it to switch.
Ideally within the unit test I would like to be able to mimic the switching of `StoreGraphs` within the `DualGraph` instance. I Initialised a DualGraph instance in an attempt to get switching to occur in test cases but still could not replicate the bug even while testing it on the current master which happens to have the bug still active in that branch.
If you have any ideas on this I'm all ears - If i cannot get it to replicate in a unit test should we remove the unit test as it is redundant? @arcturus2

arcturus2 replied:
@aldebaran30701 i really like the idea of writing a unit test when fixing bugs as it helps detect it should it come back. I think you are mostly there with it and I'd like to help out to see if I can give some guidance to get the graph switching part to work. I'll get back to you.

aldebaran30701 replied:
I've branched off master and added the unit test back in as to replicate the bug and work out how to get the graphs switching in the unit test. https://github.com/constellation-app/constellation/tree/layersSyncTest
I've tried accessing the instances from within the `LockingStoreGraph` but ran into many issues with private variables

I have just tried to get two writable graphs - both grabbed instance 1. Which is not what is needed.

aldebaran30701 replied:
>I'd like to help out to see if I can give some guidance to get the graph switching part to work


@arcturus2 I think I'll take you up on that offer - Above is the branch I've been testing with.

arcturus2 replied:
@aldebaran30701 please see PR #625 and let me know if this helps or not.

arcturus2 commented:
@aldebaran30701 are you ready for this to be merged in?

aldebaran30701 commented:
@arcturus2 Go right ahead.

arcturus2 commented:
@aldebaran30701 ta.

@antares1470 and @cygnus-x-1 would both of you be able to review this PR when you get a chance? I will do the same.

This a a change to the `StoreGraph` so a decent amount of testing is required.

arcturus2 commented:
@aldebaran30701 i did some random testing and I was able to add some nodes to layer 2 and 3 and then I proceeded to turn on and off layer 2 very quickly. After some time enabling/disabling layer 2 did not do anything.

aldebaran30701 commented:
> @aldebaran30701 i did some random testing and I was able to add some nodes to layer 2 and 3 and then I proceeded to turn on and off layer 2 very quickly. After some time enabling/disabling layer 2 did not do anything.

Hi @arcturus2 I haven't been able to replicate this on my machine yet, but I may be doing something wrong. I made a new graph, added a few vertices and then manually added a few vertices to layers 2 and 3. I switched on layer 2 really fast, switched on layer 3 really fast, and also tried 2-3-2-3-2-3 etc. also tried with layer 3 on, selecting 2 fast, and vice versa. 

This most recent change rounds the plugins which set the layer queries and the layer bit mask into one plugin - which should help alleviate the issue above if it is related to that. 

To test it, I had to bring in #637 so that the context menu would work again. 
Would you be able to either test this wrt the original issue, or let me know how I should be trying to reproduce it and I can verify it is fixed.

arcturus2 commented:
This is really interesting @aldebaran30701. After you highlighted the issue with the context menus not working when I tried this morning I wasn't seeing nodes added to layers. So I wasn't sure how the layer 2 was originally disappearing/appearing but then stopped to do so all together.

Let me test this again and get back to you.

arcturus2 commented:
@aldebaran30701 so I manually added the layer because the context menu isn't working (which should be resolved with a merge of master into this branch) and it all seems to be working fine. Would you be able to merge master in before we merge this PR?

aldebaran30701 commented:
@arcturus2 Perfect. I've updated from the current master so should be all good to go 👍
